### PR TITLE
Sort out mm sources which are architecture-independent 

### PR DESCRIPTION
**Pull request by maroueneboubakri**
To ease the port of OP-TEE OS to more architectures, we try to make maximum reuse of the existing sources. With regards to memory management:

- `core_memprot.h` does not define any architecture-dependent definition and could be moved from `core/arch/arm/include/mm/` to `core/include/mm/`.
- `core_mmu.h` is used by `core_memprot.h` and used by sources which are located outside arch tree either in core (such as `dt.c` and `tee_ta_manager.c`) or in `libutils` `(ftrace.c`).  A suggestion is to move it also to `core/include/mm/` but let the architecture implement the functions prototyped there. Defintions which are arch specific could still be there for now. 
- `mobj*` use definitions form `core_mmu.h`. Moving `core_mmu.h` to `core/include/mm/` lets pulling `mobj*` from arch tree, therefore, can be reused by any arch.
- `tee_pager.h` could be also moved to `core/include/mm/` and let the arch implement the functions defined there.
- `tee_mm.c` can be also moved to core/mm since it is architecture-independent and `tee_mm.h` is already in core/include/mm.

Signed-off-by: Marouene Boubakri <marouene.boubakri@nxp.com>

### CODE REVIEW

File: core/kernel/pm.c [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #include <kernel/pm.h>
 #include <mm/core_memprot.h>
 #include <string.h>
+#include <malloc.h>
```
Comment (jenswi-linaro) [inline core/kernel/pm.c:11]: Unrelated change?

Comment (maroueneboubakri) [reply core/kernel/pm.c:11]: Compiler complains about `realloc`

Comment (jenswi-linaro) [reply core/kernel/pm.c:11]: Please put this in a separate commit in this PR. It may seem a bit picky but it improves the history.

Comment (jforissier) [reply core/kernel/pm.c:11]: Isn't `realloc()` in `<stdlib.h>`?

Comment (maroueneboubakri) [reply core/kernel/pm.c:11]: Agree since `<stdlib.h>` includes `<malloc.h>`

**Comment by jenswi-linaro** [comment]
The commit message isn't easy to understand. It would probably be easier if this was split in multiple commits.
One could perhaps be described as:
```
core: move core_memprot.h to core/include/mm

Moves core_memprot.h to core/include/mm since it is architecture independent.
```
and another
```
core: move enum teecore_memtypes and teecore_memtype_name() to core_memprot.h.

Moves enum teecore_memtypes and teecore_memtype_name() to core_memprot.h
to them part of architecture independent code.
```
and so on. When moving stuff around it's usually good to take it in small steps.

**Review by jforissier — COMMENTED** [review]
For commit "core: kernel: pm.c: add missing header stdlib.h":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Review by jforissier — COMMENTED** [review]
For commit "core: mm: move core_memprot.h to core/include/mm":

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/drivers/stpmic1.c [hunk 2]
```diff
@@ -587,22 +587,21 @@ bool stpmic1_regulator_is_valid(const char *name)
 	return get_regulator_data(name);
 }
 
-TEE_Result stpmic1_regulator_levels_mv(const char *name,
-				       const uint16_t **levels,
-				       size_t *levels_count)
+int stpmic1_regulator_levels_mv(const char *name,
```
Comment (jforissier) [inline core/drivers/stpmic1.c:590]: If the value is not used this should return `void`. Generally speaking error codes are preferably `TEE_Result` whenever possible.

**Review by jforissier — COMMENTED** [review]
For commmit "core: drivers: stm32_rng.c: include thread.h":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: core/drivers/stpmic1.c (hunk 2 above)
Comment (maroueneboubakri) [reply core/drivers/stpmic1.c:590]: The function implementation returns a value but indeed  the caller does not check it. Would say maybe `pmic_describe_levels`  in` scmi_server.c` should add the check.

**Comment by jenswi-linaro** [comment]
For commits "core: kernel: pm.c: add missing header stdlib.h",
"core: mm: move core_memprot.h to core/include/mm" and
"core: drivers: stm32_rng.c: include thread.h":
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/drivers/stpmic1.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stpmic1.c:590]: yes, I would prefer to lease this function as-is and test the return value at caller place.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits 
"core: kernel: pm.c: add missing header stdlib.h"
"core: mm: move core_memprot.h to core/include/mm"
"core: drivers: stm32_rng.c: include thread.h"

For commit "core: driver: stpmic1: do not use TEE_Result as return type": see my review comment.

**Comment by maroueneboubakri** [comment]
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
> "core: kernel: pm.c: add missing header stdlib.h"
> "core: mm: move core_memprot.h to core/include/mm"
> "core: drivers: stm32_rng.c: include thread.h"
> 
> For commit "core: driver: stpmic1: do not use TEE_Result as return type": see my review comment.

I would make it `void `as suggested by jforissier to be less intrusif.

**Comment by jforissier** [comment]
> > `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
> > "core: kernel: pm.c: add missing header stdlib.h"
> > "core: mm: move core_memprot.h to core/include/mm"
> > "core: drivers: stm32_rng.c: include thread.h"
> > For commit "core: driver: stpmic1: do not use TEE_Result as return type": see my review comment.
> 
> I would make it `void `as suggested by jforissier to be less intrusif.

Nits:
- In commit subjects/descriptions please use parentheses after function names: `function stpmic1_regulator_levels_mv() ...` (advantage, you can even drop the word `function` in most cases).
- Since you have made a choice to change to `void` there is no need to mention a different alternative in the description (remove the "Either ...").

**Comment by maroueneboubakri** [comment]
> > > `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
> > > "core: kernel: pm.c: add missing header stdlib.h"
> > > "core: mm: move core_memprot.h to core/include/mm"
> > > "core: drivers: stm32_rng.c: include thread.h"
> > > For commit "core: driver: stpmic1: do not use TEE_Result as return type": see my review comment.
> > 
> > 
> > I would make it `void `as suggested by jforissier to be less intrusif.
> 
> Nits:
> 
> * In commit subjects/descriptions please use parentheses after function names: `function stpmic1_regulator_levels_mv() ...` (advantage, you can even drop the word `function` in most cases).
> * Since you have made a choice to change to `void` there is no need to mention a different alternative in the description (remove the "Either ...").

Done

**Comment by jforissier** [comment]
For commit "core: driver: stpmic1: do not use TEE_Result as return type":

`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

But note: @etienne-lms I'd like to understand why you think it is better to keep the return status and check it in the caller. All checks are on device names that are defined in the source code so any error would be an internal (programming) error right? Hence `assert()` seems OK. Am I missing something?

**Comment by jenswi-linaro** [comment]
I'd prefer not adding any more commits to this PR. The last three commit is risking this PR to stall.

**Review by etienne-lms — COMMENTED** [review]
As for the 4 first commits, it looks like they could be merged.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
* "core: driver: stpmic1: do not use TEE_Result as return type"
* "core: drivers: stm32_rng.c: include thread.h"  (my r-b tag misses its `Reviewed-by: `)

**Comment by etienne-lms** [comment]
@maroueneboubakri, maybe you can create a P-R with th 4 first commits and get them merged.

**Comment by jforissier** [comment]
> @maroueneboubakri, maybe you can create a P-R with th 4 first commits and get them merged.

I suggest keeping only these 4 commits here and applying the review tags. The subsequent ones can go in another PR.

**Comment by jforissier** [comment]
@maroueneboubakri last commit needs fixing (Etienne's R-b).
